nullaway/src/main/java/com/uber/nullaway/handlers/contract/ContractCheckHandler.java

```diff
@@ -0,0 +1,120 @@
+import com.sun.source.util.TreePathScanner;
+import com.sun.tools.javac.code.Symbol;
+import com.uber.nullaway.NullAway;
+import com.uber.nullaway.Nullness;
+import com.uber.nullaway.handlers.BaseNoOpHandler;
+
+/**
+ * This Handler parses the jetbrains @Contract annotation and tries to check if the contract is
+ * followed.
+ *
+ * <p>Currently, it supports the case when there is only one clause in the contract. The clause of
+ * the form in which all the elements of the antecedent are either of "_", "null" or "!null", and
+ * the consequent is "!null" is supported. The handler checks and warns under the conditions of the
+ * antecedent if the consequent is "!null" and there is a return statement with "nullable" or "null"
+ * expression.
+ */
+public class ContractCheckHandler extends BaseNoOpHandler {
+
+  @Override
+  public void onMatchMethod(
+      NullAway analysis, MethodTree tree, VisitorState state, Symbol.MethodSymbol methodSymbol) {
+    Symbol.MethodSymbol callee = ASTHelpers.getSymbol(tree);
+    Preconditions.checkNotNull(callee);
+    // Check to see if this method has an @Contract annotation
+    String contractString = getContractFromAnnotation(callee);
+
+    if (contractString != null) {
+      // Found a contract, lets parse it.
+      String[] clauses = contractString.split(";");
+      if (clauses.length != 1) {
+        return;
+      }
+
+      String clause = clauses[0];
+      String[] antecedent =
+          getAntecedent(clause, tree, analysis, state, callee, tree.getParameters().size());
+      String consequent = getConsequent(clause, tree, analysis, state, callee);
+
+      boolean supported = true;
+
+      for (int i = 0; i < antecedent.length; ++i) {
+        String valueConstraint = antecedent[i].trim();
+        if (!(valueConstraint.equals("_")
+            || valueConstraint.equals("!null")
+            || valueConstraint.equals("null"))) {
+          supported = false;
+        }
+      }
+
+      if (!consequent.equals("!null")) {
+        supported = false;
+      }
+
+      if (!supported) {
+        return;
+      }
+
+      // we scan the method tree for the return nodes and check the contract
+      new TreePathScanner<Void, Void>() {
+        @Override
+        public Void visitReturn(ReturnTree returnTree, Void unused) {
+
+          final VisitorState returnState = state.withPath(getCurrentPath());
+          final Nullness nullness =
+              analysis
+                  .getNullnessAnalysis(returnState)
+                  .getNullnessForContractDataflow(
+                      new TreePath(returnState.getPath(), returnTree.getExpression()),
+                      returnState.context);
+
+          if (nullness == Nullness.NULLABLE || nullness == Nullness.NULL) {
+
+            final String errorMessage =
+                "Method has @Contract("
+                    + contractString
+                    + "), but this appears to be violated, as a @Nullable value may be returned.";
```

Comment:
This message should still show: method name, as well as note that `@Nullable` is returned when the pre-conditions are satisfied. e.g. 

```
Method [methodName] has @Contract("[contractString]"), but this appears to be violated, as a @Nullable value may be returned when [argument] is `@NonNull`.
```

You can just put "when contract pre-conditions are satisfied" in the case of multiple arguments being relevant, but in the common case of e.g. `_, !null, _ -> !null` it should list the argument's name.

Reply:
so by argument you mean if we have `return someExpression;` then we use `someExpression` in the message?

Reply:
Let me know how it looks now. I think I was confused because I didn't know that having only one antecedent as `!null` is a very common thing in contracts.